### PR TITLE
*:record raft cluster bootstrap time

### PR DESCRIPTION
**Pull request by nolouch**
close https://github.com/pingcap/pd/issues/641
PTAL @disksing @siddontang

### CODE REVIEW

File: server/cluster.go [hunk 1]
```diff
@@ -243,6 +261,7 @@ func checkBootstrapRequest(clusterID uint64, req *pdpb.BootstrapRequest) error {
 }
 
 func (s *Server) bootstrapCluster(req *pdpb.BootstrapRequest) (*pdpb.BootstrapResponse, error) {
+
```
Comment (overvenus) [inline server/cluster.go]: Please remove the empty line.

File: server/api/router.go [hunk 2]
```diff
@@ -46,6 +46,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	router.HandleFunc("/api/v1/schedulers/{name}", schedulerHandler.Delete).Methods("DELETE")
 
 	router.Handle("/api/v1/cluster", newClusterHandler(svr, rd)).Methods("GET")
+	router.HandleFunc("/api/v1/raft/bootstrap/time", newClusterHandler(svr, rd).GetRaftClusterBootstrapTime).Methods("GET")
```
Comment (overvenus) [inline server/api/router.go]: How about 'bootstrap_time', consistent with the key below.

**Comment by siddontang** [comment]
We should also get the PD cluster bootstrap time, not only raft TiKV cluster.

File: server/api/router.go (hunk 2 above)
Comment (siddontang) [inline server/api/router.go]: I  think /cluster/raft is ok, which will return the whole raft cluster information including the bootstrap time.

File: server/cluster.go [hunk 3]
```diff
@@ -206,6 +220,10 @@ func makeStoreKeyPrefix(clusterRootPath string) string {
 	return strings.Join([]string{clusterRootPath, "s", ""}, "/")
 }
 
+func makeBootstrapTimeKey(clusterRootPath string) string {
+	return strings.Join([]string{clusterRootPath, "bootstrap_time"}, "/")
```
Comment (disksing) [inline server/cluster.go]: How about _raft_bootstrap_time_?

Comment (nolouch) [inline server/cluster.go]: the cluster root path already include "raft" https://github.com/pingcap/pd/pull/645/files/feb470d10155b16c2144b964c8becb920e9c64f7#diff-d412b9240514b30a308e4b4c8208ae72R169

Comment (disksing) [inline server/cluster.go]: Or _tikv_bootstrap_time_?

Comment (siddontang) [inline server/cluster.go]: /cluster/raft/status

Comment (disksing) [inline server/cluster.go]: Got it.

File: server/util.go [hunk 4]
```diff
@@ -333,3 +333,16 @@ func diffRegionKeyInfo(origin *RegionInfo, other *RegionInfo) string {
 
 	return strings.Join(ret, ",")
 }
+
+func parseTimestamp(data []byte) (time.Time, error) {
+	if data == nil {
+		return zeroTime, nil
+	}
```
Comment (disksing) [inline server/util.go]: Unnecessary to check nil, _bytesToUint64_ takes good care of it.
BTW, prefer `if len(data) == 0` to `if data == nil` when check a slice.

Comment (nolouch) [inline server/util.go]: we need return zeroTime if load data not exist

**Comment by disksing** [comment]
LGTM

**Comment by nolouch** [comment]
PTAL @disksing @siddontang

File: server/api/cluster_test.go [hunk 5]
```diff
@@ -57,3 +58,16 @@ func (s *testClusterInfo) TestCluster(c *C) {
...
+
+func (s *testClusterInfo) TestGetClusterStatus(c *C) {
+	url := fmt.Sprintf("%s/cluster/status", s.urlPrefix)
+	status := server.ClusterStatus{}
+	err := readJSONWithURL(url, &status)
+	c.Assert(status.BootstrapTime, IsNil)
+	now := time.Now()
+	mustBootstrapCluster(c, s.svr)
+	err = readJSONWithURL(url, &status)
+	fmt.Printf("%+v", status)
```
Comment (disksing) [inline server/api/cluster_test.go]: Remove debug messages.

File: server/cluster.go [hunk 6]
```diff
@@ -206,6 +244,14 @@ func makeStoreKeyPrefix(clusterRootPath string) string {
 	return strings.Join([]string{clusterRootPath, "s", ""}, "/")
 }
 
+func makeRaftClusterStatusPrefix(clusterRootPath string) string {
+	return strings.Join([]string{clusterRootPath, "status"}, "/")
```
Comment (disksing) [inline server/cluster.go]: How about `path.Join()`?

File: server/cluster.go [hunk 7]
```diff
@@ -56,6 +60,13 @@ type RaftCluster struct {
 
 	wg   sync.WaitGroup
 	quit chan struct{}
+
+	status *ClusterStatus
+}
+
+// ClusterStatus saves some state information
+type ClusterStatus struct {
+	BootstrapTime *time.Time `json:"raft_bootstrap_time,omitempty"`
```
Comment (siddontang) [inline server/cluster.go]: /BootstrapTime/RaftBootstrapTime/s

File: server/util.go [hunk 8]
```diff
@@ -333,3 +333,12 @@ func diffRegionKeyInfo(origin *RegionInfo, other *RegionInfo) string {
 
 	return strings.Join(ret, ",")
 }
+
+func parseTimestamp(data []byte) (time.Time, error) {
+	nano, err := bytesToUint64(data)
```
Comment (siddontang) [inline server/util.go]: please add test to cover it.

File: server/cluster.go [hunk 9]
```diff
@@ -56,6 +60,13 @@ type RaftCluster struct {
 
 	wg   sync.WaitGroup
 	quit chan struct{}
+
+	status *ClusterStatus
+}
+
+// ClusterStatus saves some state information
+type ClusterStatus struct {
+	RaftBootstrapTime *time.Time `json:"raft_bootstrap_time,omitempty"`
```
Comment (siddontang) [inline server/cluster.go]: don't use the pointer here.

**Comment by siddontang** [comment]
PTAL @disksing

File: server/util_test.go [hunk 10]
```diff
@@ -40,3 +41,17 @@ func (s *testMinMaxSuite) TestMinDuration(c *C) {
 	c.Assert(minDuration(time.Second, time.Minute), Equals, time.Second)
 	c.Assert(minDuration(time.Second, time.Second), Equals, time.Second)
 }
+
+var _ = Suite(&testUtilSuite{})
+
+type testUtilSuite struct{}
+
+func (s *testUtilSuite) TestParseTimestap(c *C) {
+	for i := 0; i < 3; i++ {
```
Comment (siddontang) [inline server/util_test.go]: add a test to cover zero time.

File: server/util_test.go [hunk 11]
```diff
@@ -54,4 +54,8 @@ func (s *testUtilSuite) TestParseTimestap(c *C) {
 		c.Assert(err, IsNil)
 		c.Assert(nt, Equals, t)
 	}
+	data := []byte("pd")
+	nt, err := parseTimestamp(data)
+	c.Assert(err, NotNil)
+	c.Assert(nt.Equal(zeroTime), IsTrue)
```
Comment (siddontang) [inline server/util_test.go]: `Assert(nt, Equals, zeroTime)`

Comment (nolouch) [inline server/util_test.go]: we do no care time location in here. use `time.Equal` will just compare `sec` and `nsec`.

**Comment by siddontang** [comment]
LGTM

PTAL @disksing

**Comment by disksing** [comment]
LGTM.
